[WIP] Improve error reporting

urso opened the pull request:
- Introduces txfile/txerr for error generation and custom queries on errors
- Errors must be consumable by: Application error handling, Application logging, End User, Operator, and Developer
  - Application error handling requires well defined comparable error codes

An error value:
  - is a tree of errors
  - A node type can have 0, 1 or N children. txerr assumes (not enforced by types):
    - errors capturing 1 cause only to implement `Cause() error`
    - errors capturing N causes must implement `Causes() []error`
    - errors not implementing `Cause` or `Causes`, or returning `nil` are leaf nodes (root causes).
  - each node in the tree can provide additional context (accessed via Methods)
  - common error context (optional):
    - Operation: Document operation/function call. Common format: `<package>/<function name>`.
    - Kind: comparable error value:
      - must be a constant, so error kinds can be compared by value
      - use custom type (`type kind ...`), so error comparison does not produce false positives when comparing error codes from multiple packages.
    - Context: additional context specific error fields (must be a string for now)
    - Message: custom error message (Note: `vfs` package implements `Path() string`, but no Message).

The txerr package provides functions to 'build', iterate, query and print error values. Using `txerr.Is(myKind, err)`, the full error tree is searched for an error matching the error kind. With errors being nested, one can check for multiple conditions (like an 'AND' on error codes) by checking `if txerr.Is(Failure, err) && txerr.Is(RootCauseWeCanHandle, err) { ... }`

The `reason` type is a placeholder for now, forcing the compiler to generate compile time errors while doing the refactoring.

TODO:
- Introduce custom error type with additional context in packages:
  - [x] txfile/txerr package for querying errors
  - [x] txfile
  - [x] txfile/pq
  - [x] internal/vfs (not based on txfile/txerr, but integrates with txerr thanks to compatible interfaces)

ph commented on .travis.yml:
```diff
@@ -5,6 +5,5 @@ os:
   - osx
 
 go:
-  - '1.9'
```
we might want something that doesn't use `strings.builder` or shim it.

urso replied:
There are a many changes in go-txfile we will only see in 6.4. Usage in beats will be updated as well. But yeah, let's try to support 1.9 for now.

urso commented on errors.go:
```diff
@@ -1,7 +1,65 @@
+	OSOtherError          = vfs.ErrOSOther
+	OperationNotSupported = vfs.ErrNotSupported
+	LockFailed            = vfs.ErrLockFailed
+)
```
I did not name them ErrX, cause I see the `kind` more like a code or descriptive context element used by applications in order to classify the error (originally I have had another custom interface type). e.g. one will use these kinds like `if txerr.Is(txfile.PermissionError, err) { ... }` or `if txerr.Is(txfile.OutOfMemory, err)  { ... }`.

The vfs package uses `ErrX`... still some confusion on how to do this on my side.

Note, an error kind is of type `error`, because common interface of errors should be based on builtin types only (txerr should not enforce any other types). By forcing the `kind/class/code` to be type of type `error`, we enforces dev to implement `Error() string`. My original definition of kind was `type Kind interface { Describe() string }`. But this would require other packages to use the `txerr` package. Requiring `interface{}` would make the kind comparable as well, but would not enforce any of description.

urso commented on errors.go:
```diff
@@ -14,9 +72,9 @@ var (
+	// errMmapTooLarge    = errors.New("mmap too large")
 	errFileSizeTooLage = errors.New("max file size to large for this system")
-	errInvalidFileSize = errors.New("invalid file size")
+	// errInvalidFileSize = errors.New("invalid file size")
```
yep

urso commented on internal/vfs/error.go:
```diff
@@ -0,0 +1,90 @@
+		return kindStr[k]
+	}
+	return "unknown"
+}
```
it's in case someone uses `txerr.Wrap(nil)`.

urso commented on internal/vfs/error.go:
```diff
@@ -0,0 +1,90 @@
+		return "no error"
+	}
+	return ""
+}
```
ups

urso commented:
> I also think the end goal is to be able to better do error assertion in higher level code to better report
a specific error to the user. (Looking at txerr.Is()) Searching in the current code the usage of Is
doesn't return anything. (I believe this would be your Application error handling)

Right. Right now `txerr` package fullfils 2 purposes:
1. Common ways of querying errors (for use by `txfile` users). Most important is `txerr.Is`. This functionality is not used within `txfile`, but made available for `txfile` users.
2. Provide some means to construct an error.

> The thing I am bit nervous is the added complexity in defining errors. I am questioning myself, could
we solve the problem with returning concrete error directly but with a better contextual logger
for added information.

I guess that's due to txerr providing a generic error type. Which makes me bit nervous as well. As txerr reporting is based on interfaces, we can have custom errors in `txfile` top level and custom error type in the `pq` package. The `internal/vfs` package already implements its own custom type, that is compatible to `txerr` querying. I figured I would use the same error type for `txfile` and `txfile/pq`. That's why txerr provides error construction as well. Alternatives to `txerr` providing error construction would be to:
- each package provides it's own error type (duplicate definitions)
- move common error type to `txfile` top level package

ph commented:
> I guess that's due to txerr providing a generic error type. Which makes me bit nervous as well. As txerr reporting is based on interfaces, we can have custom errors in txfile top level and custom error type in the pq package. The internal/vfs package already implements its own custom type, that is compatible to txerr querying. I figured I would use the same error type for txfile and txfile/pq. That's why txerr provides error construction as well. Alternatives to txerr providing error construction would be to:

👍  In the context of writing to disk so many things can go wrong. With your explaining I'm ok to move forward with your suggested strategy.

kvch commented on opts.go:
```diff
@@ -67,8 +71,21 @@ func (o *Options) Validate() error {
+	if o.PageSize != 0 {
+		if !isPowerOf2(uint64(o.PageSize)) {
+			return txerr.Of(InvalidConfig).
+				Msgf("pageSize %v is no power of 2", o.PageSize)
```
tyop: "is not power of 2"

kvch commented:
I agree with @ph. After you explained why it's necessary and how it is superior compared to other solution, I am on board.

kvch commented:
What is going to be the "idiomatic" way of creating new errors? I have seen both `txerr.Of(InvalidMetaPage).Msg("invalid magic number")` and
`Error{op: op, ctx: p.errCtx(), cause: cause}`.

urso commented:
> What is going to be the "idiomatic" way of creating new errors? I have seen both `txerr.Of(InvalidMetaPage).Msg("invalid magic number")` and
`Error{op: op, ctx: p.errCtx(), cause: cause}`.

Good point. I modified txerr.Error to also allow for some context. A context is some kind of structured meta-data added to an error. Mostly for reporting. Like file name, object ID (for correlation), file offset touched and so on. Just using `txerr` I ended up with a many `txerr.Op(...).Of(...).Msgf("k1=%v %k2=%v : my error message", v1, v2)`. Loads of duplicate logic, always adding in some context. So instead of using format strings I decided to create custom error types with additional context. These custom error types implement `Op, Kind, Context, Message, Cause`, and are therefor fully compatible to txerr it's generic error.

Next to adding a context I wanted to be able to 'merge' context and/or removing duplicate information. For example in the `pq` package, if an error and it's cause report the same page ID, the page ID will be removed from the causing error. This way we print:
```
pq/my-op : id=0x12345 page=10 : readable message
  pq/my-cause-op: readable message
  pq/yet-another-nested-op: with another error message
```
instead of 
```
pq/my-op: id=0x12345 page=10: readable message
  pq/my-cause-op: id=0x12345 page=10: readable message
  pq/yet-another-nested-op: id=0x12345 page=10: with another error message
```

The 'context' idea is some kind of structured logging, yet I didn't want to build a generic 'context'. Adding contexts is a pattern I added rather lately.

The finaly conventions/ideas i did apply in this PR:
- errors are interfaces
  - we can use custom error type or just txerr to build an error at will
  - simple root-cause errors without extra context are often build using `txerr.Op/Of/Msg`.
  - errors with contexts should be build with custom package error types
    - custom error types must be compatible to `txerr.Error`
- minimum common interface for all errors generated by txfile is defined by `txerr.Error`:
```
type Error interface {
	error

	Op() string
	Kind() error
	Message() string
	Cause() error
}
```
- all return types in `txerr.Error` should be builtin types. Do not enforce packages to reference txerr for custom types/interfaces
- An additional `Context() string` method is optional (I didn't enforce what a context is, as long as it's printable).
- error Kinds use type `error`, but must be uniquely comparable. That is due to interface comparison rules every package defining an error kind must define a custom type + constants for every supported error kind. Common pattern
```
type errKind int // this type declaration ensures comparisons with values from other packets will not get us false positives

//go:generate stringer -type=errKind -linecomment=true
const (
  Value1 errKind = iota // readable error message
  Value2                // another readable error message
)

func (k errKind) Error() string { // satisfy error interface
  return k.String()
}
```
- error types follow this pattern (we do not expose fields, but use methods, such that errors can always be handled as interfaces):
```
type Error struct {
  op string
  kind error
  cause error
  ctx errorCtx
  msg string
}

type errorCtx struct {
  k1 int
  k2 uint

  // optional context field (if you are too lazy to create a many error types)
  isK3 bool
  k3 string
}

// fullfill txerr.Error interface:
func (e *Error) Error() string   { return txerr.Report(e) } // implement `error` with help of txerr
func (e *Error) Op() string      { return e.op }
func (e *Error) Kind() error     { return e.kind }
func (e *Error) Cause() error    { return e.cause }
func (e *Error) Context() string { return e.ctx.String() }
func (e *Error) Message() string { return e.msg }

// context based accessors (optional)
func (e *Error) K1() int { return e.ctx.k1 }

func (ctx *errorCtx) String() string {
	buf := &strbld.Builder{}
    buf.Fmt("k1=%v k2=%v", ctx.k1, ctx.k2)
    if ctx.isK3 {
      buf.Fmt(" k3=%v", ctx.k3)
    }
	return buf.String()
}

// optional error build helpers (e.g. for wrapping another error):
func (e *Error) of(kind errKind) *Error {
  e.kind = kind
  return e
}

func (e *Error) report(s string) *Error {
  e.msg = s
  return e
}

func (e *Error) causedBy(cause error) *Error {
  e.cause = cause
  if other, ok := cause.(*Error); ok {
    mergeCtx(e, other)
  }
  return e
}

func mergeCtx(err, cause *Error) { ... }
```
- objects carying context provide functions/helpers to create errors
```
type myObj struct {...}

func (obj *myObj) methodThatFails() error {
  const op = "<package-name>/obj-<readable-op>"

  ...

  // on failure
  if err != nil {
    return obj.errWrap(op, err).of(ForApplicationUse).report("custom user message")
  }

  ...
}

func (obj *myObj) errWrap(op string, cause error) *Error {
  return obj.err(op).causedBy(cause)
}

func (obj *myObj) err(op string) *Error {
  return &Error{op: op, ctx: errCtx()}
}

func (obj *myObj) errCtx() errorCtx {
  ...
}
```


Indeed, there is quite some boilerplate code required for errors. But we just define custom errors with error contexts in 2 places. But we get nice stack-trace-like reports + error context for correlating errors and error-kinds provide us some flexible ways of dealing with error values. E.g.:
- did a write fail because out of disk space? Let's back-off and try again
- did a write fail because all file pages have been allocated -> wait for ACKer to free pages
- file creation failed? Trace of errors gives details which operation exactly failed and why.

So I used the `txerr.Op/Of/Msg` methods for building errors without context or if not enough context is available yet. As I have added these errors before introducing the custom error types, I just didn't remove the usage of txerr based error values. 

Would you prefer to remove the txerr based error builder, so it becomes more obvious how errors should be constructed?

ruflin commented on txerr/txerr.go:
```diff
@@ -0,0 +1,95 @@
+
+package txerr
+
+type Error interface {
```
Could use some docs on why these 4 methods. Should we unleash the hound on this repo? :-)

urso replied:
Will add docs to the txerr package once it's more settled. I changed some concepts like 3 times so far.

ruflin commented on file.go:
```diff
@@ -251,29 +264,45 @@ func (f *File) Close() error {
 	lock.Lock()
 	tracef("init new transaction (readonly: %v)\n", settings.Readonly)
-	tx := newTx(f, lock, settings)
+	tx := newTx(f, f.txids.Inc(), lock, settings)
```
A bit surprised that `newTx` can't return an error but checked the code and looks good.

ruflin commented on file.go:
```diff
@@ -612,11 +690,14 @@ func growFile(f *File, opts Options) error {
 // actual file size before truncating. Truncation is postponed to later transactions.
-func shrinkFile(f *File, opts Options) error {
+func shrinkFile(f *File, opts Options) reason {
+	const op = "txfile/shrink"
```
Instead of defining all these const locally would it make sense to have them together in one table and only used there? Would make it easier to see what `ops` are there.

urso replied:
Idea of `op` is to have something a programmer can grep for and immediately find the position of the function. Introducing another file with definitions will add some more indirection.

The `op` should follow the function name. That's also why there is a newline after `op`. My hope is `op` will be updated once the function name changes.

Also `op` should not occur twice (even when using a const). If we really make this a more general pattern, some linter checking for duplicates would be helpful.

ruflin commented on file.go:
```diff
@@ -720,12 +809,16 @@ func initTxReleaseRegions(f *File) {
-func withInitTx(f *File, fn func(tx *Tx) error) error {
-	tx := f.Begin()
+func withInitTx(f *File, fn func(tx *Tx) reason) reason {
+	tx, err := f.beginTx(TxOptions{Readonly: false})
```
Seems like a change in behaviour? Looks good to me as I think it's a improvement but wanted to mention it.

urso replied:
Yeah, I noticed Begin did not return an error. But it might be illegal to start a new transaction (file closed, readonly transaction). As I was improving error reporting/handling, I introduced this missing error.

ruflin commented on internal/strbld/strbld_old.go:
```diff
@@ -0,0 +1,34 @@
+// specific language governing permissions and limitations
+// under the License.
+
+// +build !go1.10
```
What about Go 1.11?

ruflin commented on pq/error.go:
```diff
@@ -17,16 +17,116 @@
+	txerr.Error
+}
+
+type Error struct {
```
Does each package need it's own definition of `Error`? They are not 100% identical but look very similar.

ruflin commented on pq/pq.go:
```diff
@@ -71,53 +82,44 @@ func MakeRoot() [SzRoot]byte {
+	}
+
+	// use pointer address as ID for correlating error messages
+	q.id = queueID(uintptr(unsafe.Pointer(q)))
```
Can you share some more background here on why and how you correlate error messages?

urso replied:
It's mostly for debugging purposes.

Each queue has it's own ID (in this case based on pointer). Transaction IDs are an uint64 increasing counter. E.g. in the context of transaction a many errors can occur. Each error generated within the same transaction will print the `txid`. Once a txid is known, one can search for all other related messages by txid.
In the future we might have multiple queues per spool file (e.g. due to versioning). Using the queue ID we can tell if one or multiple queues fail from time to time + one can correlate multiple errors to one queue. Would be nice if we can set the queue id from the outside and store queue ids in the spool file.

ruflin commented on pq/pq.go:
```diff
@@ -71,53 +82,44 @@ func MakeRoot() [SzRoot]byte {
 	tracef("open queue: %p (pageSize: %v)\n", q, pageSize)
 	traceQueueHeader(root)
-
-	tail := q.accessor.ParsePosition(&root.tail)
```
If I get it right this was now move to the Writer.

urso replied:
yeah. queue accessors are lazily instantiated by now.

ruflin commented on pq/pq.go:
```diff
@@ -127,14 +129,32 @@ func (q *Queue) Close() error {
-	tx := q.accessor.BeginRead()
+	tx, err := q.accessor.BeginRead()
+	if err != nil {
+		return -1
```
What about return also an `error` as second argument? Would give a bit more context on what went wrong.

ruflin commented on pq/pq.go:
```diff
@@ -154,25 +174,56 @@ func (q *Queue) Pending() int {
+	flushed := q.settings.Flushed
+	writer, err := newWriter(&q.accessor, q.pagePool, writeBuffer, tail, flushed)
+	if err != nil {
+		panic("TODO")
```
?

urso replied:
oh, nice one. Need to grep for more TODOs :)

ruflin commented on pq/reader.go:
```diff
@@ -49,39 +49,61 @@ func newReader(accessor *access) (*Reader, error) {
 // User must execute Done, to close the file transaction.
-func (r *Reader) Begin() {
+func (r *Reader) Begin() error {
+	const op = "txfile/reader-begin"
```
Why is this prefixed by `txfile` and below the op is prefixed by `pq`?

urso replied:
ups. Maybe I should use this pattern:

```
const pkg = "pq/"

func (r *Reader) Begin() error {
  const op = pkg+"reader-begin"
  ...
}
```

ruflin commented on pq/util.go:
```diff
@@ -28,23 +28,14 @@ func getPage(tx *txfile.Tx, id txfile.PageID) ([]byte, error) {
-		page = pool.NewPageWith(id, buf)
-		return nil
-	})
+	fn(page)
```
Did not fully understand why `fn` does not return an error anymore.

urso replied:
We have 2 error types. Plus there was only one function ever reporting an error. This function does not return an error anymore, but stores the error in a local variable. By not returning an error anymore and capturing it I can distinguish where the error actually occured + I don't have to deal with different error types.

ruflin commented on tx.go:
```diff
@@ -152,24 +155,39 @@ func (tx *Tx) RootPage() (*Page, error) {
 		return nil, nil
 	}
-	return tx.Page(tx.rootID)
+	return tx.getPage("txfile/tx-access-root", tx.rootID)
```
Should this use a const?

urso replied:
👍  yeah, I got a little lazy at times...

ruflin commented on tx.go:
```diff
@@ -571,8 +600,9 @@ func (tx *Tx) rollbackChanges() error {
-		return err
+		// getting file size failed. State is valid, but we can not truncate
+		// ¯\_(ツ)_/¯
+		return
```
Still sounds like an error to me worth noting?

urso replied:
It's not really an error in this location. The file and txfile state is still valid, but the OS failed us when returning bytes to the filesystem. This does not affect txfile from operating correctly. Next transaction will try to truncate again, until we succeed at some point in time.

kvch commented:
@urso Thanks for the detailed explanation. I would prefer to remove it if possible.

urso commented:
@kvch I removed error construction via txerr.

urso commented:
Applied some more changes to better support zap (as used in beats) based logging. It's a little painfull and not fully consistent.
When logging via zap one must use `%+v` for multiline format strings. Using Infow, zap will only log the first line and causes are written into an error field:

Sample error + logging:

```
	err := &Error{
		op: "op/test",
		cause: &Error{
			op:  "op/nested",
			ctx: errorCtx{a: "valueA", b: "valueB"},
			msg: "ups",
			cause: &Error{
				op:   "op/more nesting",
				ctx:  errorCtx{b: "anotherB"},
				kind: errors.New("nested twice"),
			},
		},
		ctx: errorCtx{a: "anotherValue"},
		msg: "message 1",
	}

	logp.Info("using logp.Info with v: %v", err)
	logp.Info("using logp.Info with +v: %+v", err)
	logp.Info("using logp.Info with s: %s", err) // `s` formatter does use Error(), not String()

	logger := logp.NewLogger("test")
	logger.Info("using logger.Info:", err)
	logger.Infof("using logger.Infof with v: %v", err)
	logger.Infof("using logger.Infof with +v: %+v", err)
	logger.Infow("using logger.Infow", "error", err)
```

sample output:

```
2018-07-17T15:33:00.267+0200    INFO    zaptest/main.go:72    using logp.Info with v: op/test: key1='anotherValue': message 1
2018-07-17T15:33:00.268+0200    INFO    zaptest/main.go:73    using logp.Info with +v: op/test: key1='anotherValue': message 1
    op/nested: key1='valueA' key2='valueB': ups
    op/more nesting: key2='anotherB': nested twice
2018-07-17T15:33:00.268+0200    INFO    zaptest/main.go:74    using logp.Info with s: op/test: key1='anotherValue': message 1
2018-07-17T15:33:00.268+0200    INFO    [test]    zaptest/main.go:77    using logger.Info:op/test: key1='anotherValue': message 1
2018-07-17T15:33:00.268+0200    INFO    [test]    zaptest/main.go:78    using logger.Infof with v: op/test: key1='anotherValue': message 1
2018-07-17T15:33:00.268+0200    INFO    [test]    zaptest/main.go:79    using logger.Infof with +v: op/test: key1='anotherValue': message 1
    op/nested: key1='valueA' key2='valueB': ups
    op/more nesting: key2='anotherB': nested twice
2018-07-17T15:33:00.268+0200    INFO    [test]    zaptest/main.go:80    using logger.Infow    {"error": "op/test: key1='anotherValue': message 1", "errorCauses": [{"error": "op/nested: key1='valueA' key2='valueB': ups", "errorCauses": [{"error": "op/more nesting: key2='anotherB': nested twice", "errorCauses": []}]}]}
```

Note, the key-value fields are still included in the error message. They are not extracted. For supporting extacting of context fields we would need to add zap as a direct dependency to go-txfile (which I would not like to do).

kvch commented:
I like the improvements you've made to logging. The multiline indentation looks a bit weird, but at least it's going to be easy to find errors when you use "eyegrep" on the log.

Theoretical question: what happens when we switch to a new logging lib? The formatting should be adjusted again to make sure errors provide sufficient information?

ph commented:
I did another round of review of the code, it look pretty cool can't wait to use it more in the core.
I would have preferred to stay away from multiline log output but it is much clearer that way vs _InfoW_ 👍

urso commented:
I will create smaller PRs from the changes applied here and add some more godoc where appropriate. This PR will be closed, once all smaller once are merged.

- [ ] Add internal/strbld package
- [ ] Unified OS error reporting in internal/vfs
- [ ] Add helper package txerr
- [ ] Update error reporting in txfile top-level package
- [ ] Update error reporting in txfile/pq package

urso commented:
PR has been split into multiple 'smaller' PRs, that also including some more cleanups. Closing.
